tests/conftest.py

```diff
@@ -0,0 +1,181 @@
+np_only = pytest.mark.skipif(not np_backend(), reason="Test for numpy backend only.")
+torch_only = pytest.mark.skipif(
+    not pytorch_backend(), reason="Test for pytorch backends only."
+)
+tf_only = pytest.mark.skipif(
+    not tf_backend(), reason="Test for tensorflow backends only."
+)
+
+np_and_tf_only = pytest.mark.skipif(
+    not (np_backend() or tf_backend()),
+    reason="Test for numpy and tensorflow backends only.",
+)
+np_and_torch_only = pytest.mark.skipif(
+    not (np_backend() or pytorch_backend()),
+    reason="Test for numpy and pytorch backends only.",
+)
+np_and_autograd_only = pytest.mark.skipif(
+    not (np_backend() or autograd_backend()),
+    reason="Test for numpy and autograd backends only.",
+)
+autograd_and_torch_only = pytest.mark.skipif(
+    not (autograd_backend() or pytorch_backend()),
+    reason="Test for autograd and torch backends only.",
+)
+autograd_and_tf_only = pytest.mark.skipif(
+    not (autograd_backend() or tf_backend()),
+    reason="Test for autograd and tf backends only.",
+)
+
+np_autograd_and_tf_only = pytest.mark.skipif(
+    not (np_backend() or autograd_backend() or tf_backend()),
+    reason="Test for numpy, autograd and tensorflow backends only.",
+)
+np_autograd_and_torch_only = pytest.mark.skipif(
+    not (np_backend() or autograd_backend() or pytorch_backend()),
+    reason="Test for numpy, autograd and pytorch backends only.",
+)
+autograd_tf_and_torch_only = pytest.mark.skipif(
+    np_backend(), reason="Test for backends with automatic differentiation only."
+)
+
+
+def pytorch_error_msg(a, b, rtol, atol):
+    msg = f"\ntensor 1\n{a}\ntensor 2\n{b}"
+    if torch.is_tensor(a) and torch.is_tensor(b):
+        if a.dtype == torch.bool and b.dtype == torch.bool:
+            diff = torch.logical_xor(a, b)
+            msg = msg + f"\ndifference \n{diff}"
+        else:
+            diff = torch.abs(a - b)
+            msg = msg + f"\ndifference \n{diff}\nrtol {rtol}\natol {atol}"
+    return msg
+
+
+class TestData:
+    """Class for TestData objects."""
+
+    def generate_tests(self, smoke_test_data, random_test_data=[]):
+        """Wraps test data with corresponding markers.
+
+        Parameters:
+            smoke_test_data : list
+                Test data that will be marked as smoke
+
+            random_test_data : list
+                Test data that will be marked as random.
+                Optional, default: []
+
+        Returns:
+            _: list
+                Tests.
+        """
+        smoke_tests = [
+            pytest.param(*data.values(), marks=smoke) for data in smoke_test_data
+        ]
+        random_tests = [pytest.param(*data, marks=random) for data in random_test_data]
+        return smoke_tests + random_tests
+
+
+class Parametrizer(type):
+    """Metaclass for test files.
+
+    Parametrizer decorates every function inside the class with pytest.mark.parametrizer
+    (except class methods and static methods). Two conventions have to followed
```

Comment:
NIT: Small edits on the English syntax:

Two conventions need to be respected:

1) There should be a TestData object named 'testing_data'.
2) Every test function should have its corresponding data function inside TestData object. Ex:
      test_exp() should have method called exp_data() inside 'testing_data.'"""